internal/core/docker_context.go

```diff
@@ -0,0 +1,328 @@
+func (s *metadataStore) list() ([]metadata, error) {
+	ctxDirs, err := listRecursivelyMetadataDirs(s.root)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	res := make([]metadata, 0, len(ctxDirs))
+	for _, dir := range ctxDirs {
+		c, err := s.getByID(contextdir(dir))
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return nil, fmt.Errorf("read metadata: %w", err)
+		}
+		res = append(res, c)
+	}
+
+	return res, nil
+}
+
+type contextdir string
+
+func isContextDir(path string) bool {
+	s, err := os.Stat(filepath.Join(path, metaFile))
+	if err != nil {
+		return false
+	}
+
+	return !s.IsDir()
+}
+
+func listRecursivelyMetadataDirs(root string) ([]string, error) {
+	fis, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []string
+	for _, fi := range fis {
+		if fi.IsDir() {
+			if isContextDir(filepath.Join(root, fi.Name())) {
+				result = append(result, fi.Name())
+			}
+
+			subs, err := listRecursivelyMetadataDirs(filepath.Join(root, fi.Name()))
+			if err != nil {
+				return nil, err
+			}
+
+			for _, s := range subs {
+				result = append(result, filepath.Join(fi.Name(), s))
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func parseTypedOrMap(payload []byte, getter typeGetter) (any, error) {
+	if len(payload) == 0 || string(payload) == "null" {
+		return nil, nil
+	}
+
+	if getter == nil {
+		var res map[string]any
+		if err := json.Unmarshal(payload, &res); err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+
+	typed := getter()
+	if err := json.Unmarshal(payload, typed); err != nil {
+		return nil, err
+	}
+
+	return reflect.ValueOf(typed).Elem().Interface(), nil
+}
+
+// getHomeDir returns the home directory of the current user with the help of
+// environment variables depending on the target operating system.
+// Returned path should be used with "path/filepath" to form new paths.
+//
+// On non-Windows platforms, it falls back to nss lookups, if the home
+// directory cannot be obtained from environment-variables.
+//
+// If linking statically with cgo enabled against glibc, ensure the
+// osusergo build tag is used.
+//
+// If needing to do nss lookups, do not disable cgo or set osusergo.
+//
+// getHomeDir is a copy of [pkg/homedir.Get] to prevent adding docker/docker
+// as dependency for consumers that only need to read the config-file.
+//
+// [pkg/homedir.Get]: https://pkg.go.dev/github.com/docker/docker@v26.1.4+incompatible/pkg/homedir#Get
+func getHomeDir() string {
+	home, _ := os.UserHomeDir()
+	if home == "" && runtime.GOOS != "windows" {
+		if u, err := user.Current(); err == nil {
+			return u.HomeDir
+		}
+	}
+	return home
+}
+
+// configurationDir returns the directory the configuration file is stored in
+func configurationDir() string {
+	configDir := os.Getenv(envOverrideConfigDir)
+	if configDir == "" {
+		configDir = filepath.Join(getHomeDir(), configFileDir)
```

Comment:
suggestion: just return, avoiding the unnecessary assignment and arguably clearer to the reader.
```suggestion
		return filepath.Join(getHomeDir(), configFileDir)
```

Reply:
Indeed. Copied using a sync.Once, but I removed it for simplicity. Changed